### ISSUE TITLE
False positive for ImportOrder for default IntelliJ Idea settings

### ISSUE THREAD

**Issue by ychulovskyy** [checkstyle/checkstyle#3398]
/Test.java

```
import data.Class;
import data.*;

public class Test {
    public static void main(String[] args) {
        Object o1 = new Class();
        Object o2 = new SomeClass();
    }
}
```

/data/Class.java

```
package data;

public class Class {
}
```

/data/SomeClass.java

```
package data;

public class SomeClass {
}
```

/checkstyle.xml

```
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
        "-//Puppy Crawl//DTD Check Configuration 1.3//EN"
        "http://www.puppycrawl.com/dtds/configuration_1_3.dtd">

<module name="Checker">
    <module name="TreeWalker">

        <!-- Checks for imports                              -->
        <!-- See http://checkstyle.sf.net/config_import.html -->
        <module name="ImportOrder">
            <property name="groups" value="*,javax,java"/>
            <property name="ordered" value="true"/>
            <property name="separated" value="false"/>
            <property name="option" value="bottom"/>
            <property name="sortStaticImportsAlphabetically" value="true"/>
        </module>
    </module>
</module>
```

/$ java -Duser.language=en -cp checkstyle-7.1-all.jar com.puppycrawl.tools.checkstyle.Main -c checkstyle.xml src/Test.java

```
Starting audit...
[ERROR] C:\Projects\CheckstyleImportOrderIssue\src\Test.java:2: Wrong order for
'data.*' import. [ImportOrder]
Audit done.
Checkstyle ends with 1 errors.
```

Expected behavior:
There should be no errors to make Checkstyle compatible with default IntelliJ Idea settings.
The problem is that Checkstyle checks that imports are in alphabetical order, which is correct.
But '*' is before [a-z], that is why Checkstyle expects the following import order:

```
import data.*;
import data.Class;
```

but Intellij Idea puts puts \* after more specific classes:

```
import data.Class;
import data.*;
```

which makes sense, because developers would like to see specific classes before *

Expected implementation:
allow such order by default (if it doesn't conflict with other IDE)
or add a configuration attribute to allow such ordering

P.S.: Why the hell we should import data.Class when we already imported data.*?
It's because java.lang.* (and java.lang.Class) imported implicitly.
If we don't import data.Class then compiler assumes that we refer to java.lang.Class.
And we cannot rename the class "Class" because it's in external library
##

**Comment by romani** [comment]
thanks a lot.

is it a regression of 7.1 release ?
Did it work before for you ?

**Comment by ychulovskyy** [comment]
I'm introducing Checkstyle into new code base.
But looks like it's not a regression because I can reproduce the same issue with checkstyle-6.11.2

**Comment by romani** [comment]
@linelect ,

> allow such order by default (if it doesn't conflict with other IDE)

If you want to resolve this issue , please investigate conflicts in other IDEs (Eclipse, NetBeans) and we need to try to update Check without any new property. If not possible .... lets discuss.

**Comment by linelect** [comment]
Hello, i done some investigation, and got strange results:

Intellij IDEA before organize imports:
```
import com.company.zata.*;
import com.company.data.*;
import com.company.data.Byte;
import com.company.data.Class;
import com.company.zata.Long;
import com.company.zata.Short;
```

Intellij IDEA after organize imports:
```
import com.company.data.Byte;
import com.company.data.Class;
**import com.company.data.*;
import com.company.zata.*;**
import com.company.zata.Long;
import com.company.zata.Short;
```


Eclipse before organize imports:
```
import com.company.data.Byte;
import com.company.data.Class;
import com.company.zata.Long;
import com.company.zata.Short;
import com.company.data.*;
import com.company.zata.*;
```

Eclipse after organize imports:
```
import com.company.data.*;
import com.company.data.Byte;
import com.company.data.Class;
import com.company.zata.*;
import com.company.zata.Long;
import com.company.zata.Short;
```

NetBeans before organize imports:
```
import com.company.data.Byte;
import com.company.data.Class;
import com.company.zata.Long;
import com.company.zata.Short;
import com.company.data.*;
import com.company.zata.*;
```

NetBeans after organize imports:
```
import com.company.data.Byte;
import com.company.data.Class;
**import com.company.data.*;
import com.company.zata.*;**
import com.company.zata.Long;
import com.company.zata.Short;
```

Eclipse sorting imports alphabetical.  
How sorting imports Intellij IDEA and NetBeans - I do not understand.
What do you think?

**Comment by ychulovskyy** [comment]
Thanks a lot @linelect !
It's up to @romani how to handle this.
IMHO, an option that allows to ignore order for imports with "*" is OK.

**Comment by romani** [comment]
@ychulovskyy , the problem right now is that we do not understand how Idea works.
We can not mimic its behavior.
In examples above : 
```
import com.company.data.Byte;
import com.company.data.Class;
import com.company.data.*;
import com.company.zata.*;
import com.company.zata.Long;
import com.company.zata.Short;
```

How `.*` could appear close to each other ?

**Comment by ychulovskyy** [comment]
@romany & @linelect 
I've tried to organize imports in latest IntelliJ IDEA 2016.2.4
Build #IC-162.2032.8, built on September 9, 2016
JRE: 1.8.0_45-b14 amd64
JVM: Java HotSpot(TM) 64-Bit Server VM by Oracle Corporation

Before organizing import:
````
import com.company.zata.*;
import com.company.data.*;
import com.company.data.Byte;
import com.company.data.Class;
import com.company.zata.Long;
import com.company.zata.Short;
````

After organizing import:
````
import com.company.data.Byte;
import com.company.data.Class;
import com.company.data.*;
import com.company.zata.Long;
import com.company.zata.Short;
import com.company.zata.*;
````

Looks predictable.
See Main class from https://github.com/ychulovskyy/ImportExample

**Comment by romani** [comment]
@ychulovskyy , thanks a lot, your results looks reasonable.
My Idea(the same version) works the same as yours.  
@linelect , please share your version of Idea, and try example project that @ychulovskyy prepared. Please share results.

**Comment by romani** [comment]
as we all got the same results, looks like we need to name option as "onDemandImportLastInGroup"
https://docs.oracle.com/javase/specs/jls/se8/html/jls-7.html#jls-7.5.2 and we need to report bugs to Eclipse and Netbeans to support such sorting.

**Comment by linelect** [comment]
I got result with Intellij IDEA 15.0.1, but Eclipce and Netbeans is latest version.

My Intellij IDEA 15.0.1 for example project that @ychulovskyy prepared got result:
```
import com.company.data.*;
import com.company.zata.*;
```

**Comment by romani** [comment]
@linelect 
Please create bug/feaure-request on Eclipse to support that ordering. 
Please create bug on Netbeans to resolve that problem.
Please proceed with implementation in checkstyle.

**Comment by linelect** [comment]
@ychulovskyy ,
I done some investigation for static imports:

IntelliJ IDEA 2016.2.4 before:
```
import static com.company.data.Class.*;
import static com.company.data.Class.Byte;
import static com.company.zata.Class2.Long;
import static com.company.zata.Class2.*;
```

IntelliJ IDEA 2016.2.4 after:
```
import static com.company.data.Class.Byte;
import static com.company.data.Class.*;
import static com.company.zata.Class2.*;
import static com.company.zata.Class2.Long;
```

IntelliJ IDEA 2016.2.5 before:
```
import static com.company.data.Class.*;
import static com.company.data.Class.Byte;
import static com.company.zata.Class2.Long;
import static com.company.zata.Class2.*;
```

IntelliJ IDEA 2016.2.5 after:
```
import static com.company.data.Class.*;
import static com.company.data.Class.Byte;
import static com.company.zata.Class2.*;
import static com.company.zata.Class2.Long;
```

Eclipse Neon.1a Release (4.6.1) and NetBeans 8.2 before:
```
import static com.company.zata.Class2.Long;
import static com.company.zata.Class2.*;
import static com.company.data.Class.*;
import static com.company.data.Class.Byte;
```

Eclipse Neon.1a Release (4.6.1) and NetBeans 8.2 after: (option Number of static imports needed for "*" = 1):
```
import static com.company.data.Class.*;
import static com.company.zata.Class2.*;
```

Eclipse Neon.1a Release (4.6.1) and NetBeans 8.2 after :(option Number of static imports needed for "*" = 3):
```
import static com.company.data.Class.Byte;
import static com.company.data.Class.Field2;
import static com.company.zata.Class2.Field4;
import static com.company.zata.Class2.Long;
```

I prepared example project - https://github.com/linelect/TestStaticImportsOrder.git

So, last version Intellij IDEA (2016.2.5) wrong sorting static imports.

**Comment by romani** [comment]
@linelect ,
if no reply explanatory reply from ychulovskyy , proceed with issues creation:

1) against IntelijIdea for static imports with "*". Behavior should be the same as in normal imports.

2) 
> Please create bug/feaure-request on Eclipse to support that ordering.

Do ordering the same as IntelijIdea by special option to let the same team work in Eclipse and Intelij without changing order of imports.

3)
> Please create bug on Netbeans to resolve that problem.

Inconsistent ordering of "*".

------------------

Till we find work of IDEs consistent, we can not proceed.

**Comment by ychulovskyy** [comment]
@linelect thanks for testing. I have the same result for IntelliJ IDEA.
Basing on this, there is nothing we can do except submit an issue, as @romani suggested.

**Comment by rnveach** [comment]
@ychulovskyy @linelect Just following up, was an issue ever filed?
Can we have the link to it?

**Comment by ychulovskyy** [comment]
@rnveach Are you asking if I submitted an issue for IntelliJ Idea to describe that static and non static import orders are different? No, I didn't.
However original issue here is that I cannot setup Checkstyle to be compatible with default IntelliJ Idea settings for non static import order.
Are there any plans to fix this?

**Comment by romani** [comment]
issue as created by @linelect -  https://youtrack.jetbrains.com/issueMobile/IDEA-163103

**Comment by rnveach** [comment]
@romani issue tracker now gives 404.

Is there anything else to be done in this issue?

**Comment by romani** [comment]
Link should be https://youtrack.jetbrains.com/issue/IDEA-163103

It was fixed and released at "2018.3".

**Comment by romani** [comment]
I am closing issue, if smth is still missing, create new issue.
